api/v1alpha1/circuitbreaker_types.go

```diff
@@ -0,0 +1,50 @@
+// Copyright Envoy Gateway Authors
+// SPDX-License-Identifier: Apache-2.0
+// The full text of the Apache license is available in the LICENSE file at
+// the root of the repo.
+
+package v1alpha1
+
+// CircuitBreakers defines the Circuit Breakers configuration.
+type CircuitBreakers struct {
+	// List of Circuit Breaker Thresholds
+	// At most one Thresholds resource is supported.
+	//
+	// +kubebuilder:validation:MaxItems:=1
+	// +optional
+	Thresholds []Thresholds `json:"thresholds,omitempty"`
+}
+
+type Thresholds struct {
+	// The maximum number of connections that Envoy will establish to the referenced backend (per xRoute).
+	//
+	// +kubebuilder:validation:Minimum=0
+	// +kubebuilder:validation:Maximum=4294967295
+	// +kubebuilder:default=1024
+	// +optional
+	MaxConnections *uint32 `json:"maxConnections,omitempty"`
+
+	// The maximum number of pending requests that Envoy will queue to the referenced backend (per xRoute).
+	//
+	// +kubebuilder:validation:Minimum=0
+	// +kubebuilder:validation:Maximum=4294967295
+	// +kubebuilder:default=1024
+	// +optional
+	MaxPendingRequests *uint32 `json:"maxPendingRequests,omitempty"`
+
+	// The maximum number of parallel requests that Envoy will make to the referenced backend (per xRoute).
+	//
+	// +kubebuilder:validation:Minimum=0
+	// +kubebuilder:validation:Maximum=4294967295
+	// +kubebuilder:default=1024
+	// +optional
+	MaxRequests *uint32 `json:"maxParallelRequests,omitempty"`
+
+	// The maximum number of parallel retries that Envoy will allow to the referenced backend (per xRoute).
+	//
+	// +kubebuilder:validation:Minimum=0
+	// +kubebuilder:validation:Maximum=4294967295
+	// +kubebuilder:default=3
+	// +optional
+	MaxRetries *uint32 `json:"maxRetries,omitempty"`
```

Comment:
`MaxRetries` is part of the retry and would be better included in the retry, the corresponding field is `maxParallel` #2168

Reply:
I believe EG should distinguish the `MaxConcurrentRetries` of a cluster and the `MaxRetries` of an individual request.

`MaxConcurrentRetries` belongs to the Circuit Breaker configuration, and `MaxRetries` belongs to the Retry configuration.

Reply:
I agree. However, note that https://github.com/envoyproxy/gateway/pull/2168 deals with `RetryBudget` which is also a part of Envoy's Circuit Breaker settings. 

In Envoy, the separation of route and cluster settings is pretty clear. Multiple routes can point to the same cluster, and so `MaxRetries`, `RetryBudget` will apply to all the traffic coming from routes that share an upstream cluster. The motivation is to protect the upstream system from a retry storm. 

In Envoy Gateway, we have a cluster for each `xRoute`. So, it doesn't make much difference if these settings are managed under the `Retries` or the `CircuitBreakers` section. It is important that the users understand the implications of these settings - overflowing retries will be queued and later dropped.  

If in the future Envoy Gateway does support a notion of shared backends (e.g. by translating services to clusters in some situations) and Envoy Gateway will support a `SharedBackendTrafficPolicy`, I expect that this policy will include `CircuitBreakers` but not `Retries`. So, for future reusability, it could be better to have these settings under the circuit breaker types.  

I'm willing to drop `MaxRetries` from this PR for now. We can continue the discussion in https://github.com/envoyproxy/gateway/pull/2168 on the best location for these settings in the API and implement it as part of that PR. WDYT?

Reply:
IMO, the concurrent max retries setting belongs to Circuit Breaker logic because it enforces back pressure on the clients. Therefore, EG probably should not mix it with the request retries configuration.

Reply:
I think we should distinguish the difference between `circuit_breaker` and `retryStrategy` functions for users, and should not be limited to the design of envoy. Although  `retry budget` and `concurrent max retries` are implemented in `circuit_breaker` in envoy, for users, they are all retry functions( provides richer uses for the retry function).

Shared cluster is indeed a point that needs to be considered, but now I am not sure how to use it. Maybe it is an implementation similar to istio DestinationRule resource. If so, you can patch the BackendTrafficPolicy to dr.

Reply:
> IMO, we should distinguish between the `circuit_breaker` and `retryStrategy` functions for users, as they offer distinct features, not limited to the design of Envoy. Even though `retry budget` and `concurrent max retries` are implemented in the `circuit_breaker` in Envoy, for users, these encompass retry functions that provide richer options for retry operations.

I vote -1 on this.

 Even though both have "retries" in their name, they serve two different purposes. The concurrent max retries setting is inherently associated with the Circuit Breaker, which fails requests quickly when a lot of retries happen and apply back pressure on downstream. On the other hand, request retries are specifically designed to mitigate transient network issues. Would love more insights from @kflynn and other @envoyproxy/gateway-maintainers

Reply:
Since @arkodg https://github.com/envoyproxy/gateway/pull/2284#discussion_r1428347907 and @tmsnan support removing `MaxRetries` from this API proposal, I'll go ahead and remove it. If we eventually decide that `CircuitBreakers` should contain these settings, we can add them later on.